### PR TITLE
[Metadata] Several bugfixes

### PR DESCRIPTION
**Pull request by valentinEmpy**
…services

When someone undeploys an MTA WITHOUT using the --delete-services CLI option, the mta
metadata must be removed from the service in order to avoid future
detection of the same MTA.
Currently, it uses a workaround by setting empty ("") mta_id in
services' metadata because the cf-java-client has a bug which does not
allow deleting metadata labels/annotations.

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/detect/DeployedMtaDetector.java [hunk 1]
```diff
@@ -44,10 +46,21 @@
         List<CloudEntity> mtaMetadataEntities = mtaMetadataEntityCollectors.stream()
                                                                            .map(collector -> collector.collect(client, criteria))
                                                                            .flatMap(List::stream)
+                                                                           .filter(this::nonEmptyMtaId)
                                                                            .collect(Collectors.toList());
         return mtaMetadataEntityAggregator.aggregate(mtaMetadataEntities);
     }
 
+    private boolean nonEmptyMtaId(CloudEntity entity) {
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/detect/DeployedMtaDetector.java]: Let's add unit test for that

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/detect/DeployedMtaDetector.java [hunk 2]
```diff
@@ -44,10 +46,21 @@
...
+        Metadata serviceMetadata = entity.getV3Metadata();
+        if (serviceMetadata == null || serviceMetadata.getLabels() == null) {
+            return false;
+        }
+        String mtaId = serviceMetadata.getLabels()
+                                      .get(MtaMetadataLabels.MTA_ID);
+        return !StringUtils.isEmpty(mtaId);
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/detect/DeployedMtaDetector.java]: Is it a good idea to use Optional for such composition of nullable objects

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetachServicesFromMtaStep.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+import javax.inject.Named;
+import java.text.MessageFormat;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+@Named("detachServicesFromMtaStep")
+@Scope(BeanDefinition.SCOPE_PROTOTYPE)
+public class DetachServicesFromMtaStep extends SyncFlowableStep {
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetachServicesFromMtaStep.java]: Can we have some unit tests for the step

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java [hunk 4]
```diff
@@ -117,18 +116,10 @@ protected CloudApplicationExtended getApplication(Module module) {
                                                 .dockerInfo(parseParameters(parametersList, new DockerInfoParser()))
                                                 .attributesUpdateStrategy(getApplicationAttributesUpdateStrategy(parametersList))
                                                 .v3Metadata(ApplicationMetadataBuilder.build(deploymentDescriptor, module,
-                                                                                             resourcesAndResourceTypes))
+                                                                                             getApplicationServices(module)))
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java]: What will happen if resource has type "existing-service-key"

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BuildCloudUndeployModelStep.java [hunk 5]
```diff
@@ -144,20 +154,23 @@ private void setComponentsToUndeploy(DelegateExecution context, List<String> ser
...
 
-    private boolean shouldDeleteService(List<DeployedMtaApplication> appsToKeep, String service, Set<String> servicesForApplications) {
+    private boolean shouldDeleteService(String service, List<DeployedMtaApplication> appsToKeep, Set<String> servicesForApplications,
+                                        List<String> servicesForCurrentDeployment) {
         return appsToKeep.stream()
                          .flatMap(module -> module.getBoundMtaServices()
                                                   .stream())
                          .noneMatch(service::equalsIgnoreCase)
-            && !servicesForApplications.contains(service);
+            && !servicesForApplications.contains(service) && !servicesForCurrentDeployment.contains(service);
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BuildCloudUndeployModelStep.java]: Add unit test

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetachServicesFromMtaStep.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                                                     .getGuid();
+            client.updateServiceMetadata(serviceGuid, getMetadataWithEmptyMtaId(serviceMetadata));
+        }
+    }
+
+    // FIXME Once the fix for deleting metadata is available in cf-java-client, the whole MTA metadata must be deleted instead of setting an empty MTA_ID value
+    private Metadata getMetadataWithEmptyMtaId(Metadata metadata) {
+        return Metadata.builder()
+                       .from(metadata)
+                       .label(MtaMetadataLabels.MTA_ID, StringUtils.EMPTY)
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetachServicesFromMtaStep.java]: Better clean everything MTA related, not only MTA_ID

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java (hunk 4 above)
Comment (valentinEmpy) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java]: Tested, seems fine as the filtering works.

**Pull request merged**